### PR TITLE
Add a script and a dockerfile to build RPM pkgs for amd64 & arm64

### PR DESCRIPTION
**Pull request by dysinger**
This commit adds the ability to easily produce RPMs for Omnigres on any capable machine with Docker installed. Please review the ./pkgs/mkpkgs script for notes on setup. Then just call ./pkgs/mkpkgs at the command line. If you only want one archiecture then call `ARCH=linux/arm64 ./pkgs/mkpkgs`.

Please provide as much criticism as you like. I'll fixup commits and rebase.

The build machine setup for this has been exercising this script. Try it yourself if you like.

### CODE REVIEW

File: pkgs/rocky.dockerfile [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+        --output-type rpm \
+        --name $pkg \
+        --version $vers \
+        --iteration $ITERATION \
+        --prefix /usr/pgsql-$POSTGRES_VER/ \
+        --chdir $tmp \
+        $deps \
+        . ;\
+done
+# NOTE THAT IT WOULD BE BETTER TO HAVE --depends 'xyz >= 1.0' FOR EACH
```
Comment (yrashk) [inline pkgs/rocky.dockerfile:91]: I was able to solve it with the following patch:

```diff
diff --git a/pkgs/rocky.dockerfile b/pkgs/rocky.dockerfile
index d635082..6f3d461 100644
--- a/pkgs/rocky.dockerfile
+++ b/pkgs/rocky.dockerfile
@@ -65,19 +65,23 @@ for pkg in $(cat /build/artifacts.txt|awk -F'=' '{print $1}') ; do \
     deps="--depends postgresql$POSTGRES_VER" ;\
     for d in $(echo $pkgline|awk -F'#' '{print $2}'|tr ',' ' ') ; do \
         n=$(echo $d|awk -F'=' '{print $1}') ;\
-        _=$(echo $d|awk -F'=' '{print $2}') ;\
+        v=$(echo $d|awk -F'=' '{print $2}') ;\
         case $n in \
             dblink|pgcrypto) n="postgresql$POSTGRES_VER-contrib"   ;;\
             plpy*)           n="postgresql$POSTGRES_VER-plpython3" ;;\
         esac ;\
-        deps="$deps --depends $n" ;\
+        if [ "$v" = "*" ]; then \
+          deps="$deps --depends '$n'" ;\
+        else \
+          deps="$deps --depends '$n = $v'" ;\
+        fi \
     done ;\
     tmp=$(mktemp -d) ;\
     mkdir -p "$tmp/lib" ;\
     mkdir -p "$tmp/share/extension" ;\
     cp -a /build/packaged/$pkg--*.so        "$tmp/lib/"             2>/dev/null || true ;\
     cp -a /build/packaged/extension/$pkg--* "$tmp/share/extension/" 2>/dev/null || true ;\
-    ${GEM_HOME}/bin/fpm \
+    eval ${GEM_HOME}/bin/fpm \
         --input-type dir \
         --output-type rpm \
         --name $pkg \
```

File: pkgs/mkpkgs [hunk 2]
```diff
@@ -0,0 +1,48 @@
+#!/usr/bin/env bash
+
+# NOTE: you'll need to create a arm64 buildx container first
```
Comment (yrashk) [inline pkgs/mkpkgs:3]: Looks like I didn't need to do this on my macOS setup

File: pkgs/mkpkgs [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+source "${DIR}/versions.txt"
+
+# BULID FOR REDHAT.
+for arch in $(echo $ARCH|tr ',' ' '); do
+    # NOTE: I have this building architectures sequentially because I'm on a
+    # laptop. On a beefy CI machine, it might be better to do this in one pass
+    # maxing the machine with parallel builds like the following example:
+    # `docker buildx build --platform linux/amd64,linux/arm64 --push`
+    ROCKY_IMAGE_NAME=omnigres:rocky-pgdg${POSTGRES_VER}-$(echo $arch|sed "s/linux\///")-$REVISION
+    ROCKY_CONTAINER_NAME=$(cat /proc/*/*/*/uuid||uuidgen) ;# WORKS ON LINUX & MACOS
```
Comment (yrashk) [inline pkgs/mkpkgs:32]: nitpick: may be use $arch, $POSTGRES_VER and other relevant details for the name? Easier to debug if things go wrong

File: pkgs/rocky.dockerfile [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+ENV PATH /usr/pgsql-${POSTGRES_VER}/bin:${PATH}
+RUN cmake -DPG_CONFIG=$(which pg_config) -DOPENSSL_CONFIGURED=1 /omni
+RUN make -j all
+RUN make package_extensions
+
+WORKDIR /pkgs
+
+# PACKAGING
+RUN \
+for pkg in $(cat /build/artifacts.txt|awk -F'=' '{print $1}') ; do \
```
Comment (yrashk) [inline pkgs/rocky.dockerfile:62]: Worth extracting this to a shell script to avoid all these trailing "\"?

File: pkgs/mkpkgs [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+DIR=$(git rev-parse --show-toplevel $(dirname $0)|head -n 1)
+
+# ITERATION FOR THE PACKAGES
+ITERATION=${ITERATION:-1}
+
+# GIT REVISION OF SOURCE
+REVISION=${REVISION:-"$(git rev-parse --short HEAD)"}
+
+# READ VERSIONS
+source "${DIR}/versions.txt"
```
Comment (yrashk) [inline pkgs/mkpkgs:23]: Do we need this anywhere?

**Review by yrashk — COMMENTED** [review]
Thank you for this amazing effort! It's a great start!

Some things I can think of:

* Minor: details such as license, description, URL, etc. We can address this separately.


---

We may have a case where we merge a dependency A of B, but the latest version of B hasn't been changed. Ideally, we should prevent this during CI. Our build system will otherwise pick up a new version of A and populate artifacts.txt accordingly. This will produce a new RPM with a new version of the dependency. 

It's a broader question to discuss, not necessarily part of this packaging effort.

File: pkgs/rocky.dockerfile (hunk 1 above)
Comment (dysinger) [reply pkgs/rocky.dockerfile:91]: I had the if/case there to set the version of the dependency. It barfs on the build machine. Where are you trying this? This is not working on the latest fpm on rocky linux.

File: pkgs/mkpkgs (hunk 3 above)
Comment (dysinger) [reply pkgs/mkpkgs:32]: We can name things here however you like. It's not important to the output of the rpms.

File: pkgs/rocky.dockerfile (hunk 4 above)
Comment (dysinger) [reply pkgs/rocky.dockerfile:62]: Could. It's so small though. You decide.

File: pkgs/mkpkgs (hunk 5 above)
Comment (dysinger) [reply pkgs/mkpkgs:23]: It might have been leftover from a different tact. Let me check.

File: pkgs/rocky.dockerfile (hunk 1 above)
Comment (dysinger) [reply pkgs/rocky.dockerfile:91]: This is the battle I had with fpm and the reason for the NOTE at the bottom.

Comment (dysinger) [reply pkgs/rocky.dockerfile:91]: Oh maybe your eval is helping it behave. I'll give it a try.

File: pkgs/mkpkgs (hunk 3 above)
Comment (yrashk) [reply pkgs/mkpkgs:32]: That's why it is a nitpick :)

File: pkgs/rocky.dockerfile (hunk 4 above)
Comment (yrashk) [reply pkgs/rocky.dockerfile:62]: I find writing stuff with trailing backslashes is error-prone. Harder to debug. Not instrumental to the functionality.

File: pkgs/rocky.dockerfile (hunk 1 above)
Comment (yrashk) [reply pkgs/rocky.dockerfile:91]: It absolutely did work here. Did you try applying the patch?

The crux of the solution is in adding `eval`

File: pkgs/mkpkgs (hunk 2 above)
Comment (yrashk) [reply pkgs/mkpkgs:3]: We have both types of CI build machines

Comment (dysinger) [reply pkgs/mkpkgs:3]: I'll be more specific.

File: pkgs/mkpkgs (hunk 5 above)
Comment (dysinger) [reply pkgs/mkpkgs:23]: deleted.

File: pkgs/rocky.dockerfile (hunk 1 above)
Comment (yrashk) [reply pkgs/rocky.dockerfile]: Worked like this here:

```

#24 [18/19] RUN for pkg in $(cat /build/artifacts.txt|awk -F'=' '{print $1}') ; do     pkgline=$(grep "^$pkg\=" /build/artifacts.txt) ;    vers=$(echo $pkgline|cut -d'#' -f1|awk -F'=' '{print $2}') ;    deps="--depends postgresql16" ;    for d in $(echo $pkgline|awk -F'#' '{print $2}'|tr ',' ' ') ; do         n=$(echo $d|awk -F'=' '{print $1}') ;        v=$(echo $d|awk -F'=' '{print $2}') ;        case $n in             dblink|pgcrypto) n="postgresql16-contrib"   ;;            plpy*)           n="postgresql16-plpython3" ;;        esac ;        if [ "$v" = "*" ]; then           deps="$deps --depends '$n'" ;        else           deps="$deps --depends '$n = $v'" ;        fi     done ;    tmp=$(mktemp -d) ;    mkdir -p "$tmp/lib" ;    mkdir -p "$tmp/share/extension" ;    cp -a /build/packaged/$pkg--*.so        "$tmp/lib/"             2>/dev/null || true ;    cp -a /build/packaged/extension/$pkg--* "$tmp/share/extension/" 2>/dev/null || true ;    eval /usr/local/bin/fpm         --input-type dir         --output-type rpm         --name $pkg         --version $vers         --iteration 1         --prefix /usr/pgsql-16/         --chdir $tmp         $deps         . ;done
#24 0.284 {:timestamp=>"2024-08-21T15:37:08.411426+0000", :message=>"Created package", :path=>"omni-0.1.4-1.aarch64.rpm"}
#24 0.449 {:timestamp=>"2024-08-21T15:37:08.576818+0000", :message=>"Created package", :path=>"omni_http-0.1.0-1.aarch64.rpm"}
#24 0.637 {:timestamp=>"2024-08-21T15:37:08.765529+0000", :message=>"Created package", :path=>"omni_types-0.1.1-1.aarch64.rpm"}
#24 0.839 {:timestamp=>"2024-08-21T15:37:08.967073+0000", :message=>"Created package", :path=>"omni_sql-0.3.2-1.aarch64.rpm"}
#24 1.625 {:timestamp=>"2024-08-21T15:37:09.753040+0000", :message=>"Created package", :path=>"omni_httpc-0.1.1-1.aarch64.rpm"}
#24 1.837 {:timestamp=>"2024-08-21T15:37:09.965224+0000", :message=>"Created package", :path=>"omni_xml-0.1.1-1.aarch64.rpm"}
#24 2.019 {:timestamp=>"2024-08-21T15:37:10.146610+0000", :message=>"Created package", :path=>"omni_web-0.1.1-1.aarch64.rpm"}
#24 2.195 {:timestamp=>"2024-08-21T15:37:10.322648+0000", :message=>"Created package", :path=>"omni_aws-0.1.0-1.aarch64.rpm"}
#24 2.374 {:timestamp=>"2024-08-21T15:37:10.501933+0000", :message=>"Created package", :path=>"omni_containers-0.1.0-1.aarch64.rpm"}
#24 3.186 {:timestamp=>"2024-08-21T15:37:11.312926+0000", :message=>"Created package", :path=>"omni_httpd-0.1.2-1.aarch64.rpm"}
#24 3.351 {:timestamp=>"2024-08-21T15:37:11.478811+0000", :message=>"Created package", :path=>"omni_id-0.2.0-1.aarch64.rpm"}
#24 3.519 {:timestamp=>"2024-08-21T15:37:11.647603+0000", :message=>"Created package", :path=>"omni_json-0.1.0-1.aarch64.rpm"}
#24 3.685 {:timestamp=>"2024-08-21T15:37:11.813491+0000", :message=>"Created package", :path=>"omni_manifest-0.1.0-1.aarch64.rpm"}
#24 3.874 {:timestamp=>"2024-08-21T15:37:12.002646+0000", :message=>"Created package", :path=>"omni_mimetypes-0.1.0-1.aarch64.rpm"}
#24 4.044 {:timestamp=>"2024-08-21T15:37:12.172499+0000", :message=>"Created package", :path=>"omni_os-0.1.0-1.aarch64.rpm"}
#24 4.209 {:timestamp=>"2024-08-21T15:37:12.337573+0000", :message=>"Created package", :path=>"omni_python-0.1.0-1.aarch64.rpm"}
#24 4.376 {:timestamp=>"2024-08-21T15:37:12.504139+0000", :message=>"Created package", :path=>"omni_vfs_types_v1-0.1.0-1.aarch64.rpm"}
#24 4.552 {:timestamp=>"2024-08-21T15:37:12.679507+0000", :message=>"Created package", :path=>"omni_vfs-0.1.1-1.aarch64.rpm"}
#24 4.725 {:timestamp=>"2024-08-21T15:37:12.853684+0000", :message=>"Created package", :path=>"omni_schema-0.2.0-1.aarch64.rpm"}
#24 4.906 {:timestamp=>"2024-08-21T15:37:13.034665+0000", :message=>"Created package", :path=>"omni_seq-0.1.0-1.aarch64.rpm"}
#24 5.075 {:timestamp=>"2024-08-21T15:37:13.202571+0000", :message=>"Created package", :path=>"omni_var-0.2.1-1.aarch64.rpm"}
#24 5.244 {:timestamp=>"2024-08-21T15:37:13.372122+0000", :message=>"Created package", :path=>"omni_txn-0.2.0-1.aarch64.rpm"}
```

**Comment by dysinger** [comment]
I pushed detailed fixup commits for the things that work on the build machine.

image

**Comment by dysinger** [comment]
Adding any sort of --depends 'xyz >= 1.0' or --depends 'xyz = 1.0' barfs on the build machine. eval or not. How did you get that to work @yrashk ?

image

File: pkgs/rocky.dockerfile (hunk 4 above)
Comment (dysinger) [reply pkgs/rocky.dockerfile:62]: Ok. It doesn't bother me for small stuff, but we can move it aside.

**Comment by dysinger** [comment]
Everything is resolved except explicit --depend versioning @yrashk. This needs --autosquash on a rebase before merge or similar. lmk

File: pkgs/rocky.dockerfile (hunk 1 above)
Comment (dysinger) [reply pkgs/rocky.dockerfile]: This is the last thing I need to resolve. Trying again on the build machine.

Comment (dysinger) [reply pkgs/rocky.dockerfile]: I think it's working now. Fingers crossed and knock on wood for one final run on the build box.

**Comment by dysinger** [comment]
@yrashk all of your comments have been fixed up and the build machine is happy 
image

**Comment by yrashk** [comment]
> @yrashk all of your comments have been fixed up and the build machine is happy image

What was the problem with non-operationality of `eval` before?

**Comment by dysinger** [comment]
Maybe so. I wish I was sitting in an office with people sometimes. Going it alone and not thinking of all the possibilities sometimes gets you stuck thinking it's something else. :)

**Comment by yrashk** [comment]
Note: extensions need their `extension.control` (unversioned) files:

```
postgres=# create extension omni_httpd cascade;
ERROR:  extension "omni_httpd" is not available
DETAIL:  Could not open extension control file "/usr/pgsql-16/share/extension/omni_httpd.control": No such file or directory.
HINT:  The extension must first be installed on the system where PostgreSQL is running.
```

**Comment by dysinger** [comment]
Fixed up and pushing/testing now.

**Comment by dysinger** [comment]
image

**Comment by dysinger** [comment]
I will push one more cleanup tweak and --autosquash these fixup commits into one commit. I'm not certain what the preference is with omnigres workflow. I like to show fixup commits to address comments. But then fixup commits are great for --autosquash after the fact.

**Comment by dysinger** [comment]
"Mitte sciurum"

**Comment by dysinger** [comment]
I found one stray "ARG ARCH" in the dockerfile that doesn't hurt anything because it's not used. I will clean that up while I'm doing Ubuntu.

**Pull request merged**